### PR TITLE
AC Sensitivity Analysis

### PR DESCRIPTION
**Pull request by Djazouli**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Allow computation of all sensitivity factors in AC sensitivity analysis, along with contingencies management.

**What is the current behavior?** *(You can also link to an open issue here)*
Currently, only the BranchFlowPerInjection without compensation, without contingencies is computed

### CODE REVIEW

File: src/main/java/com/powsybl/openloadflow/sensi/AbstractSensitivityAnalysis.java [hunk 1]
```diff
@@ -79,4 +95,485 @@ protected JacobianMatrix createJacobianMatrix(EquationSystem equationSystem, Vol
         equationSystem.updateEquations(x);
         return new JacobianMatrix(equationSystem, matrixFactory);
     }
+
+    protected void fillRhsSensitivityVariable(LfNetwork lfNetwork, EquationSystem equationSystem, List<SensitivityFactorGroup> factorGroups, Matrix rhs) {
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/sensi/AbstractSensitivityAnalysis.java:99]: These three functions should be moved in order to have a more understandable ranking in the file. The ranking could be simply the calling order.

File: src/main/java/com/powsybl/openloadflow/sensi/AbstractSensitivityAnalysis.java [hunk 2]
```diff
@@ -79,4 +95,485 @@ protected JacobianMatrix createJacobianMatrix(EquationSystem equationSystem, Vol
...
+        List<ParticipatingElement> participatingElements =  step.getParticipatingElements(lfNetwork).stream().filter(filter::apply).collect(Collectors.toList());
+        ParticipatingElement.normalizeParticipationFactors(participatingElements, "bus");
+        return participatingElements;
+    }
+
+    protected List<ParticipatingElement> getParticipatingElements(LfNetwork lfNetwork, LoadFlowParameters loadFlowParameters, OpenLoadFlowParameters openLoadFlowParameters) {
+        return getParticipatingElements(lfNetwork, loadFlowParameters, openLoadFlowParameters, element -> true);
+    }
+
+    static class LfSensitivityFactor<T extends EquationTerm> {
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/sensi/AbstractSensitivityAnalysis.java:116]: Indeed this is a bit weird but it works. @floriand-e2r is it the good way to handle that?

File: src/main/java/com/powsybl/openloadflow/sensi/AcSensitivityAnalysis.java [hunk 3]
```diff
@@ -40,104 +43,97 @@ public AcSensitivityAnalysis(MatrixFactory matrixFactory) {
...
-            if (factor instanceof BranchFlowPerInjectionIncrease) {
-                BranchFlowPerInjectionIncrease injectionIncrease = (BranchFlowPerInjectionIncrease) factor;
+    protected void setReferenceActivePowerFlows(List<LfSensitivityFactor<ClosedBranchSide1ActiveFlowEquationTerm>> factors) {
+        for (LfSensitivityFactor factor : factors) {
+            factor.setFunctionReference(factor.getFunctionLfBranch().getP1());
+        }
+    }
 
-                LfBranch branch = lfNetwork.getBranchById(injectionIncrease.getFunction().getBranchId());
+    private List<SensitivityValue> getSensitivitiesContingency(List<LfSensitivityFactor<ClosedBranchSide1ActiveFlowEquationTerm>> lfFactors,
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/sensi/AcSensitivityAnalysis.java:93]: `getPostContingencySensitivityValues`?

File: src/main/java/com/powsybl/openloadflow/sensi/AcSensitivityAnalysis.java [hunk 4]
```diff
@@ -147,26 +143,117 @@ private DenseMatrix initRhs(List<SensitivityFactor> factors, LfNetwork lfNetwork
...
-        // create AC equation system
-        EquationSystem equationSystem = AcEquationSystem.create(lfNetwork, new VariableSet());
+            List<SensitivityValue> baseValues = new ArrayList<>();
 
-        // create jacobian matrix from current network state
-        VoltageInitializer voltageInitializer = getVoltageInitializer(lfParameters);
+            // we make the assumption that we ran a loadflow before, and thus this jacobian is the right one
+            try (JacobianMatrix j = createJacobianMatrix(equationSystem, new PreviousValueVoltageInitializer())) {
+                // otherwise, defining the rhs matrix will result in integer overflow
+                assert Integer.MAX_VALUE / (equationSystem.getSortedEquationsToSolve().size() * Double.BYTES) > factorGroups.size();
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/sensi/AcSensitivityAnalysis.java:190]: @geofjamg it seems that we have an issue for matrix allocation, any idea to solve this?

**Review by annetill — COMMENTED** [review]
I will make another review after your first fixes, but globally the PR looks nice!

File: src/main/java/com/powsybl/openloadflow/ac/equations/AcEquationSystem.java [hunk 5]
```diff
@@ -303,6 +305,11 @@ private static void createImpedantBranch(LfBranch branch, LfBus bus1, LfBus bus2
             equationSystem.createEquation(bus2.getNum(), EquationType.BUS_Q).addTerm(q2);
             branch.setQ2(q2);
         }
+
+        if (deriveA1 && creationParameters.isForceA1Var()) {
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/ac/equations/AcEquationSystem.java:309]: I think you can remove `creationParameters.isForceA1Var()`, no?

File: src/main/java/com/powsybl/openloadflow/util/BusState.java [hunk 6]
```diff
@@ -77,7 +79,17 @@ public void restoreDcBusState(LfBus bus) {
      * @param engine AcLoadFlowEngine to operate the PqPv switching if the bus has lost its voltage control
      */
     public static void restoreBusStates(Map<LfBus, BusState> busStates, AcloadFlowEngine engine) {
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/util/BusState.java:81]: Do we still need this function?

File: src/main/java/com/powsybl/openloadflow/sensi/AbstractSensitivityAnalysis.java [hunk 7]
```diff
@@ -7,22 +7,38 @@
...
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import java.util.Objects;
+import java.util.*;
+import java.util.function.Function;
+import java.util.stream.Collectors;
 
 /**
  * @author Geoffroy Jamgotchian <geoffroy.jamgotchian at rte-france.com>
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/sensi/AbstractSensitivityAnalysis.java:44]: @Djazouli Can you add your name as author?

File: src/main/java/com/powsybl/openloadflow/sensi/AcSensitivityAnalysis.java [hunk 8]
```diff
@@ -7,30 +7,33 @@
...
-import com.powsybl.sensitivity.factors.BranchFlowPerInjectionIncrease;
 import org.apache.commons.lang3.tuple.Pair;
 
 import java.util.*;
 import java.util.stream.Collectors;
 
-import static com.powsybl.openloadflow.OpenLoadFlowProvider.getVoltageInitializer;
-
 /**
  * @author Geoffroy Jamgotchian <geoffroy.jamgotchian at rte-france.com>
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/sensi/AcSensitivityAnalysis.java:38]: Can you add you as author?

File: src/main/java/com/powsybl/openloadflow/sensi/AcSensitivityAnalysis.java [hunk 9]
```diff
@@ -40,104 +43,97 @@ public AcSensitivityAnalysis(MatrixFactory matrixFactory) {
         super(matrixFactory);
     }
 
-    private LfBranch getLfBranch(LfNetwork lfNetwork, String branchId) {
-        LfBranch branch = lfNetwork.getBranchById(branchId);
-        if (branch == null) {
-            throw new PowsyblException("Branch '" + branchId + "' not found");
+    protected void setBaseCaseSensitivityValues(List<SensitivityFactorGroup> factorGroups, DenseMatrix factorsState) {
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/sensi/AcSensitivityAnalysis.java:46]: After reading in more detail your code, the function name is not the good one as you use it in base case sensitivity computation and in post contingency sensitivity computation.

File: src/main/java/com/powsybl/openloadflow/sensi/AbstractSensitivityAnalysis.java (hunk 2 above)
Comment (geofjamg) [reply src/main/java/com/powsybl/openloadflow/sensi/AbstractSensitivityAnalysis.java:100]: This is not ideal, as we have a lot of "raw type" warning from Sonar. Let's see that for a later refactoring.

File: src/main/java/com/powsybl/openloadflow/sensi/AcSensitivityAnalysis.java (hunk 4 above)
Comment (geofjamg) [reply src/main/java/com/powsybl/openloadflow/sensi/AcSensitivityAnalysis.java]: The solution is to solve by chunk, and chunk size should be a parameter (<= size(int). We can create a issue. for that too.

**Pull request merged**